# Part 1 - obtain, inspect, clean, and prepare data

In [1]:
# import libraries

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
sns.set()      
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)

- CSV files were downloaded from seanlahman.com

In [2]:
raw_batting = pd.read_csv('Batting.csv')
raw_fielding = pd.read_csv('Fielding.csv')
salaries = pd.read_csv('Salaries.csv')
names = pd.read_csv('People.csv')

## Data inspection

In [3]:
df = raw_batting.copy()

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110495 entries, 0 to 110494
Data columns (total 22 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   playerID  110495 non-null  object 
 1   yearID    110495 non-null  int64  
 2   stint     110495 non-null  int64  
 3   teamID    110495 non-null  object 
 4   lgID      109758 non-null  object 
 5   G         110495 non-null  int64  
 6   AB        110495 non-null  int64  
 7   R         110495 non-null  int64  
 8   H         110495 non-null  int64  
 9   2B        110495 non-null  int64  
 10  3B        110495 non-null  int64  
 11  HR        110495 non-null  int64  
 12  RBI       109739 non-null  float64
 13  SB        108127 non-null  float64
 14  CS        86954 non-null   float64
 15  BB        110495 non-null  int64  
 16  SO        108395 non-null  float64
 17  IBB       73845 non-null   float64
 18  HBP       107679 non-null  float64
 19  SH        104427 non-null  float64
 20  SF  

In [5]:
df.head()

playerID  yearID  stint teamID lgID   G   AB   R   H  2B  3B  HR   RBI  \
0  abercda01    1871      1    TRO  NaN   1    4   0   0   0   0   0   0.0   
1   addybo01    1871      1    RC1  NaN  25  118  30  32   6   0   0  13.0   
2  allisar01    1871      1    CL1  NaN  29  137  28  40   4   5   0  19.0   
3  allisdo01    1871      1    WS3  NaN  27  133  28  44  10   2   2  27.0   
4  ansonca01    1871      1    RC1  NaN  25  120  29  39  11   3   0  16.0   

    SB   CS  BB   SO  IBB  HBP  SH  SF  GIDP  
0  0.0  0.0   0  0.0  NaN  NaN NaN NaN   0.0  
1  8.0  1.0   4  0.0  NaN  NaN NaN NaN   0.0  
2  3.0  1.0   2  5.0  NaN  NaN NaN NaN   1.0  
3  1.0  1.0   0  2.0  NaN  NaN NaN NaN   0.0  
4  6.0  2.0   2  1.0  NaN  NaN NaN NaN   0.0

In [6]:
fielding_df = raw_fielding.copy()

In [7]:
fielding_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147080 entries, 0 to 147079
Data columns (total 18 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   playerID  147080 non-null  object 
 1   yearID    147080 non-null  int64  
 2   stint     147080 non-null  int64  
 3   teamID    147080 non-null  object 
 4   lgID      145568 non-null  object 
 5   POS       147080 non-null  object 
 6   G         147080 non-null  int64  
 7   GS        100924 non-null  float64
 8   InnOuts   117151 non-null  float64
 9   PO        147080 non-null  int64  
 10  A         147080 non-null  int64  
 11  E         147079 non-null  float64
 12  DP        147080 non-null  int64  
 13  PB        11835 non-null   float64
 14  WP        1169 non-null    float64
 15  SB        10079 non-null   float64
 16  CS        10079 non-null   float64
 17  ZR        1169 non-null    float64
dtypes: float64(8), int64(6), object(4)
memory usage: 20.2+ MB


In [8]:
fielding_df.head()

playerID  yearID  stint teamID lgID POS   G    GS  InnOuts  PO   A     E  \
0  abercda01    1871      1    TRO  NaN  SS   1   1.0     24.0   1   3   2.0   
1   addybo01    1871      1    RC1  NaN  2B  22  22.0    606.0  67  72  42.0   
2   addybo01    1871      1    RC1  NaN  SS   3   3.0     96.0   8  14   7.0   
3  allisar01    1871      1    CL1  NaN  2B   2   0.0     18.0   1   4   0.0   
4  allisar01    1871      1    CL1  NaN  OF  29  29.0    729.0  51   3   7.0   

   DP  PB  WP  SB  CS  ZR  
0   0 NaN NaN NaN NaN NaN  
1   5 NaN NaN NaN NaN NaN  
2   0 NaN NaN NaN NaN NaN  
3   0 NaN NaN NaN NaN NaN  
4   1 NaN NaN NaN NaN NaN

In [9]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26428 entries, 0 to 26427
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   yearID    26428 non-null  int64 
 1   teamID    26428 non-null  object
 2   lgID      26428 non-null  object
 3   playerID  26428 non-null  object
 4   salary    26428 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 1.0+ MB


In [10]:
salaries.head()

yearID teamID lgID   playerID  salary
0    1985    ATL   NL  barkele01  870000
1    1985    ATL   NL  bedrost01  550000
2    1985    ATL   NL  benedbr01  545000
3    1985    ATL   NL   campri01  633333
4    1985    ATL   NL  ceronri01  625000

In [11]:
names.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20370 entries, 0 to 20369
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   playerID      20370 non-null  object 
 1   birthYear     20259 non-null  float64
 2   birthMonth    20091 non-null  float64
 3   birthDay      19950 non-null  float64
 4   birthCountry  20311 non-null  object 
 5   birthState    19836 non-null  object 
 6   birthCity     20202 non-null  object 
 7   deathYear     10004 non-null  float64
 8   deathMonth    10003 non-null  float64
 9   deathDay      10002 non-null  float64
 10  deathCountry  9997 non-null   object 
 11  deathState    9944 non-null   object 
 12  deathCity     9991 non-null   object 
 13  nameFirst     20333 non-null  object 
 14  nameLast      20370 non-null  object 
 15  nameGiven     20333 non-null  object 
 16  weight        19554 non-null  float64
 17  height        19634 non-null  float64
 18  bats          19189 non-nu

In [12]:
names.head()

playerID  birthYear  birthMonth  birthDay birthCountry birthState  \
0  aardsda01     1981.0        12.0      27.0          USA         CO   
1  aaronha01     1934.0         2.0       5.0          USA         AL   
2  aaronto01     1939.0         8.0       5.0          USA         AL   
3   aasedo01     1954.0         9.0       8.0          USA         CA   
4   abadan01     1972.0         8.0      25.0          USA         FL   

    birthCity  deathYear  deathMonth  deathDay deathCountry deathState  \
0      Denver        NaN         NaN       NaN          NaN        NaN   
1      Mobile     2021.0         1.0      22.0          USA         GA   
2      Mobile     1984.0         8.0      16.0          USA         GA   
3      Orange        NaN         NaN       NaN          NaN        NaN   
4  Palm Beach        NaN         NaN       NaN          NaN        NaN   

  deathCity nameFirst nameLast       nameGiven  weight  height bats throws  \
0       NaN     David  Aardsma     David Allan   215.0    75.0    R      R   
1   Atlanta      Hank    Aaron     Henry Louis   180.0    72.0    R      R   
2   Atlanta    Tommie    Aaron      Tommie Lee   190.0    75.0    R      R   
3       NaN       Don     Aase  Donald William   190.0    75.0    R      R   
4       NaN      Andy     Abad   Fausto Andres   184.0    73.0    L      L   

        debut   finalGame   retroID    bbrefID  
0  2004-04-06  2015-08-23  aardd001  aardsda01  
1  1954-04-13  1976-10-03  aaroh101  aaronha01  
2  1962-04-10  1971-09-26  aarot101  aaronto01  
3  1977-07-26  1990-10-03  aased001   aasedo01  
4  2001-09-10  2006-04-13  abada001   abadan01

## Inspection analysis
- Salaries are available from only 1985 to 2016, while statistical files are available from 1871 to 2021
- Missing data is prevalent in the batting and fielding files but not the salaries file
- Each dataset has a 'playerID' that can be used for merging dataframes

## Cleaning & preparing data
- For consistency and to trim the datasets, the timeframe for all dataframes will be set to 1985-2016
- Stats for each playerID will be combined

In [13]:
# preparing batting data
df = df[(df.yearID >= 1985) & (df.yearID <= 2016)].reset_index(drop=True)

In [14]:
# combining stats
df = df.groupby(['playerID'])
df = df.sum().reset_index()

In [15]:
df.head()

playerID  yearID  stint    G   AB   R    H  2B  3B  HR   RBI   SB   CS  \
0  aardsda01   18084      9  331    4   0    0   0   0   0   0.0  0.0  0.0   
1   aasedo01   11925      6  243    5   0    0   0   0   0   0.0  0.0  0.0   
2   abadan01    6010      3   15   21   1    2   0   0   0   0.0  0.0  1.0   
3   abadfe01   16107      9  315    9   0    1   0   0   0   0.0  0.0  0.0   
4  abbotje01    9995      5  233  596  82  157  33   2  18  83.0  6.0  5.0   

   BB    SO  IBB  HBP   SH   SF  GIDP  
0   0   2.0  0.0  0.0  1.0  0.0   0.0  
1   0   3.0  0.0  0.0  0.0  0.0   0.0  
2   4   5.0  0.0  0.0  0.0  0.0   1.0  
3   0   5.0  0.0  0.0  0.0  0.0   1.0  
4  38  91.0  2.0  3.0  5.0  7.0  12.0

- The stats were summed correctly but 'yearID' and 'stint' must be dropped
- Note that only numerical columns remain after grouping by playerID

In [16]:
drop = ['yearID', 'stint']
df = df.drop(columns=drop, axis=1)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7214 entries, 0 to 7213
Data columns (total 18 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   playerID  7214 non-null   object 
 1   G         7214 non-null   int64  
 2   AB        7214 non-null   int64  
 3   R         7214 non-null   int64  
 4   H         7214 non-null   int64  
 5   2B        7214 non-null   int64  
 6   3B        7214 non-null   int64  
 7   HR        7214 non-null   int64  
 8   RBI       7214 non-null   float64
 9   SB        7214 non-null   float64
 10  CS        7214 non-null   float64
 11  BB        7214 non-null   int64  
 12  SO        7214 non-null   float64
 13  IBB       7214 non-null   float64
 14  HBP       7214 non-null   float64
 15  SH        7214 non-null   float64
 16  SF        7214 non-null   float64
 17  GIDP      7214 non-null   float64
dtypes: float64(9), int64(8), object(1)
memory usage: 1014.6+ KB


In [18]:
# preparing fielding data
fielding_df = fielding_df[(fielding_df.yearID >= 1985) & (fielding_df.yearID <= 2016)].reset_index(drop=True)

In [19]:
fielding_df.head()

playerID  yearID  stint teamID lgID POS   G    GS  InnOuts  PO   A    E  \
0   aasedo01    1985      1    BAL   AL   P  54   0.0    264.0   8  10  0.0   
1  abregjo01    1985      1    CHN   NL   P   6   5.0     72.0   1   6  1.0   
2  ackerji01    1985      1    TOR   AL   P  61   0.0    259.0  10  16  0.0   
3  adamsri02    1985      1    SFN   NL  2B   6   3.0     84.0   9  13  1.0   
4  adamsri02    1985      1    SFN   NL  3B  16  10.0    337.0   2  31  1.0   

   DP  PB  WP  SB  CS  ZR  
0   0 NaN NaN NaN NaN NaN  
1   0 NaN NaN NaN NaN NaN  
2   1 NaN NaN NaN NaN NaN  
3   1 NaN NaN NaN NaN NaN  
4   3 NaN NaN NaN NaN NaN

- Per the documentation, the last 5 columns are specific to the catcher position and will be filtered out

In [20]:
fielding_df = fielding_df.loc[:, :'DP']
fielding_df.head()

playerID  yearID  stint teamID lgID POS   G    GS  InnOuts  PO   A    E  \
0   aasedo01    1985      1    BAL   AL   P  54   0.0    264.0   8  10  0.0   
1  abregjo01    1985      1    CHN   NL   P   6   5.0     72.0   1   6  1.0   
2  ackerji01    1985      1    TOR   AL   P  61   0.0    259.0  10  16  0.0   
3  adamsri02    1985      1    SFN   NL  2B   6   3.0     84.0   9  13  1.0   
4  adamsri02    1985      1    SFN   NL  3B  16  10.0    337.0   2  31  1.0   

   DP  
0   0  
1   0  
2   1  
3   1  
4   3

- If the fielding stats were combined by playerID now the positions would be lost
- Before combining them, the most frequently played position of their career will be extracted into a new dataframe for EDA and machine learning at later stages

In [21]:
# extracting positions from fielding_df based on appearances at said positions
position_df = pd.DataFrame()

fielder_id = fielding_df.playerID.value_counts().index

# loop through fielding_df playerIDs and group positions 
for i in fielder_id:
    temp_df1 = fielding_df[fielding_df['playerID'].isin([i])]
    fielder_pos = temp_df1['POS'].value_counts().index
    temp_df2 = pd.DataFrame()
    
    # loop through positions and identify most frequently occurring position per player
    for v in fielder_pos:
        temp_df1 = temp_df1[temp_df1['POS'].isin([v])]
        temp_df2 = temp_df2.copy().append({'POS':v,'G':temp_df1.sum()['G']},ignore_index=True)
     
    # returns index (playerID) of highest game count   
    temp_pos = temp_df2['G'].idxmax()
    position_df = position_df.append({'playerID':i,'POS':temp_df2['POS'][temp_pos]},ignore_index=True)

In [22]:
position_df.head()

playerID POS
0  harrile01  3B
1  cairomi01  2B
2  bloomwi01  2B
3  hernajo01  SS
4  derosma01  2B

In [23]:
# combining stats
fielding_df = fielding_df.groupby(['playerID'])
fielding_df = fielding_df.sum().reset_index()

# drop useless columns; also dropping 'G' since it is included with batting stats
drop = ['yearID', 'stint', 'G']
fielding_df = fielding_df.drop(columns=drop, axis=1)

In [24]:
fielding_df.head()

playerID     GS  InnOuts   PO   A    E  DP
0  aardsda01    0.0   1011.0   11  29  3.0   2
1   aasedo01    0.0    965.0   22  37  1.0   2
2   abadan01    4.0    138.0   37   1  1.0   3
3   abadfe01    6.0    822.0    7  31  2.0   1
4  abbotje01  140.0   3688.0  299   2  8.0   0

In [25]:
fielding_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7177 entries, 0 to 7176
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   playerID  7177 non-null   object 
 1   GS        7177 non-null   float64
 2   InnOuts   7177 non-null   float64
 3   PO        7177 non-null   int64  
 4   A         7177 non-null   int64  
 5   E         7177 non-null   float64
 6   DP        7177 non-null   int64  
dtypes: float64(3), int64(3), object(1)
memory usage: 392.6+ KB


- Fielding data is ready for merging

In [26]:
# preparing salary data
salaries.head()

yearID teamID lgID   playerID  salary
0    1985    ATL   NL  barkele01  870000
1    1985    ATL   NL  bedrost01  550000
2    1985    ATL   NL  benedbr01  545000
3    1985    ATL   NL   campri01  633333
4    1985    ATL   NL  ceronri01  625000

In [27]:
# average salary per year (2007-2016)
avg_sal = salaries.groupby(['yearID']).mean()
avg_sal.tail(10)

salary
yearID              
2007    2.941436e+06
2008    3.136517e+06
2009    3.277647e+06
2010    3.278747e+06
2011    3.318838e+06
2012    3.458421e+06
2013    3.723344e+06
2014    3.980446e+06
2015    4.301276e+06
2016    4.396410e+06

- To account for this inflation, salaries will be adjusted based on the change in combined league salary per year, relative to combined 2016 league salary

In [28]:
# combined league salary per year
sal_yr = salaries.groupby(['yearID']).sum().reset_index()
sal_yr.head()

yearID     salary
0    1985  261964696
1    1986  307854518
2    1987  272575375
3    1988  300452424
4    1989  359995711

In [29]:
# determine change per year
sal_yr['multiplier'] = ''

for i in range(len(sal_yr)):
    
    numerator = sal_yr.iloc[-1][1]
    denominator = sal_yr.iloc[i][1]
    
    sal_yr['multiplier'][i] = numerator / denominator

In [30]:
sal_yr.head(10)

yearID     salary multiplier
0    1985  261964696   14.31543
1    1986  307854518  12.181525
2    1987  272575375  13.758166
3    1988  300452424  12.481635
4    1989  359995711  10.417172
5    1990  443881193   8.448516
6    1991  613048418   6.117196
7    1992  805543323   4.655414
8    1993  901740134   4.158778
9    1994  927836287   4.041809

In [31]:
def Adjust_Salary(full_df, multiplier_df):
    '''
    Adjusts salaries based on total change per year relative to 2016 total;
    Function uses default column names from seah lahman datasets
    '''    
    df1 = full_df
    df2 = multiplier_df

    df1['adj_salary'] = ''
    years = df1.yearID.value_counts().index


    for i in range(len(df1)):
        for year in years:
            if df1.yearID[i] == year:
                mult = df2[df2.yearID == year].multiplier
                exist_sal = df1.salary[i]
                df1['adj_salary'][i] = exist_sal * mult.values[0]

In [32]:
adj_sal = salaries.copy()

Adjust_Salary(adj_sal, sal_yr) 

In [33]:
adj_sal.adj_salary = adj_sal.adj_salary.astype('int64')

In [34]:
# combining salaries
salaries_df = adj_sal.groupby(['playerID'])
salaries_df = salaries_df.sum().reset_index()

# drop useless columns
drop = ['yearID']
salaries_df = salaries_df.drop(columns=drop, axis=1)

In [35]:
salaries_df.head()

playerID   salary  adj_salary
0  aardsda01  9259750    12772487
1   aasedo01  2300000    28499739
2   abadan01   327000      528240
3   abadfe01  3766400     4211458
4  abbotje01   985000     2301147

In [36]:
salaries_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5149 entries, 0 to 5148
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   playerID    5149 non-null   object
 1   salary      5149 non-null   int64 
 2   adj_salary  5149 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 120.8+ KB


## Merging dataframes
- Batting, fielding, and salaries will be combined into a single dataframe

In [37]:
# salaries + batting stats
df_all = pd.merge(salaries_df, df, on=['playerID'], how='right')

In [38]:
df_all.head()

playerID     salary  adj_salary    G   AB   R    H  2B  3B  HR   RBI   SB  \
0  aardsda01  9259750.0  12772487.0  331    4   0    0   0   0   0   0.0  0.0   
1   aasedo01  2300000.0  28499739.0  243    5   0    0   0   0   0   0.0  0.0   
2   abadan01   327000.0    528240.0   15   21   1    2   0   0   0   0.0  0.0   
3   abadfe01  3766400.0   4211458.0  315    9   0    1   0   0   0   0.0  0.0   
4  abbotje01   985000.0   2301147.0  233  596  82  157  33   2  18  83.0  6.0   

    CS  BB    SO  IBB  HBP   SH   SF  GIDP  
0  0.0   0   2.0  0.0  0.0  1.0  0.0   0.0  
1  0.0   0   3.0  0.0  0.0  0.0  0.0   0.0  
2  1.0   4   5.0  0.0  0.0  0.0  0.0   1.0  
3  0.0   0   5.0  0.0  0.0  0.0  0.0   1.0  
4  5.0  38  91.0  2.0  3.0  5.0  7.0  12.0

In [39]:
# batting/salaries + positions
df_all = pd.merge(position_df, df_all, on=['playerID'], how='right')

In [40]:
df_all.head()

playerID POS     salary  adj_salary    G   AB   R    H  2B  3B  HR   RBI  \
0  aardsda01   P  9259750.0  12772487.0  331    4   0    0   0   0   0   0.0   
1   aasedo01   P  2300000.0  28499739.0  243    5   0    0   0   0   0   0.0   
2   abadan01  1B   327000.0    528240.0   15   21   1    2   0   0   0   0.0   
3   abadfe01   P  3766400.0   4211458.0  315    9   0    1   0   0   0   0.0   
4  abbotje01  OF   985000.0   2301147.0  233  596  82  157  33   2  18  83.0   

    SB   CS  BB    SO  IBB  HBP   SH   SF  GIDP  
0  0.0  0.0   0   2.0  0.0  0.0  1.0  0.0   0.0  
1  0.0  0.0   0   3.0  0.0  0.0  0.0  0.0   0.0  
2  0.0  1.0   4   5.0  0.0  0.0  0.0  0.0   1.0  
3  0.0  0.0   0   5.0  0.0  0.0  0.0  0.0   1.0  
4  6.0  5.0  38  91.0  2.0  3.0  5.0  7.0  12.0

In [41]:
# batting/salaries/positions + fielding stats
df_all = pd.merge(fielding_df, df_all, on=["playerID"], how='right')

In [42]:
df_all.head()

playerID     GS  InnOuts     PO     A    E   DP POS     salary  \
0  aardsda01    0.0   1011.0   11.0  29.0  3.0  2.0   P  9259750.0   
1   aasedo01    0.0    965.0   22.0  37.0  1.0  2.0   P  2300000.0   
2   abadan01    4.0    138.0   37.0   1.0  1.0  3.0  1B   327000.0   
3   abadfe01    6.0    822.0    7.0  31.0  2.0  1.0   P  3766400.0   
4  abbotje01  140.0   3688.0  299.0   2.0  8.0  0.0  OF   985000.0   

   adj_salary    G   AB   R    H  2B  3B  HR   RBI   SB   CS  BB    SO  IBB  \
0  12772487.0  331    4   0    0   0   0   0   0.0  0.0  0.0   0   2.0  0.0   
1  28499739.0  243    5   0    0   0   0   0   0.0  0.0  0.0   0   3.0  0.0   
2    528240.0   15   21   1    2   0   0   0   0.0  0.0  1.0   4   5.0  0.0   
3   4211458.0  315    9   0    1   0   0   0   0.0  0.0  0.0   0   5.0  0.0   
4   2301147.0  233  596  82  157  33   2  18  83.0  6.0  5.0  38  91.0  2.0   

   HBP   SH   SF  GIDP  
0  0.0  1.0  0.0   0.0  
1  0.0  0.0  0.0   0.0  
2  0.0  0.0  0.0   1.0  
3  0.0  0.0  0.0   1.0  
4  3.0  5.0  7.0  12.0

- Pitchers will not be considered with regards to batting & fielding, and will be thrown out

In [43]:
df_all = df_all[df_all.POS != 'P'].reset_index(drop=True)

In [44]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3395 entries, 0 to 3394
Data columns (total 27 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   playerID    3395 non-null   object 
 1   GS          3358 non-null   float64
 2   InnOuts     3358 non-null   float64
 3   PO          3358 non-null   float64
 4   A           3358 non-null   float64
 5   E           3358 non-null   float64
 6   DP          3358 non-null   float64
 7   POS         3358 non-null   object 
 8   salary      2498 non-null   float64
 9   adj_salary  2498 non-null   float64
 10  G           3395 non-null   int64  
 11  AB          3395 non-null   int64  
 12  R           3395 non-null   int64  
 13  H           3395 non-null   int64  
 14  2B          3395 non-null   int64  
 15  3B          3395 non-null   int64  
 16  HR          3395 non-null   int64  
 17  RBI         3395 non-null   float64
 18  SB          3395 non-null   float64
 19  CS          3395 non-null  

- That's quite the discrepancy. I'm going to take a closer look at a few averages from players without salaries

In [45]:
no_sal = df_all[df_all.salary.isna()]

print('Average Games:', no_sal.G.mean().round(1), ';'
      ' Average at-bats:', no_sal.AB.mean().round(1), ';' 
      ' Average RBIs:', no_sal.RBI.mean().round(1))

Average Games: 33.0 ; Average at-bats: 71.4 ; Average RBIs: 6.9


- Considering these are career stats, the players without salaries were likely called up from the minor leagues and never received a major league contract
    - These records will be thrown out but I must be careful not to lose excess data 

In [46]:
df_all = df_all[df_all.salary.isna() == False]

In [47]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2498 entries, 0 to 3394
Data columns (total 27 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   playerID    2498 non-null   object 
 1   GS          2491 non-null   float64
 2   InnOuts     2491 non-null   float64
 3   PO          2491 non-null   float64
 4   A           2491 non-null   float64
 5   E           2491 non-null   float64
 6   DP          2491 non-null   float64
 7   POS         2491 non-null   object 
 8   salary      2498 non-null   float64
 9   adj_salary  2498 non-null   float64
 10  G           2498 non-null   int64  
 11  AB          2498 non-null   int64  
 12  R           2498 non-null   int64  
 13  H           2498 non-null   int64  
 14  2B          2498 non-null   int64  
 15  3B          2498 non-null   int64  
 16  HR          2498 non-null   int64  
 17  RBI         2498 non-null   float64
 18  SB          2498 non-null   float64
 19  CS          2498 non-null  

- There are 7 records with no defensive stats
- After researching I found that players with no position were at the tail end of their career and played only DH
- The missing positions were  reflect positions played earlier in their career

In [48]:
df_all[df_all.POS.isna()]

playerID  GS  InnOuts  PO   A   E  DP  POS     salary  adj_salary    G  \
423   burroje01 NaN      NaN NaN NaN NaN NaN  NaN   350000.0   5010400.0   86   
996    fordda01 NaN      NaN NaN NaN NaN NaN  NaN   800000.0  10705476.0   28   
1545  johnsja01 NaN      NaN NaN NaN NaN NaN  NaN   150000.0   2147314.0   17   
2015  mcraeha01 NaN      NaN NaN NaN NaN NaN  NaN  1450000.0  19370335.0  242   
2285   ortajo01 NaN      NaN NaN NaN NaN NaN  NaN   900000.0  11923629.0  237   
2952  squirmi01 NaN      NaN NaN NaN NaN NaN  NaN   200000.0   2436304.0    2   
3085  thornan01 NaN      NaN NaN NaN NaN NaN  NaN  4400000.0  58010431.0  280   

       AB    R    H  2B  3B  HR    RBI   SB   CS   BB     SO  IBB  HBP   SH  \
423   191   19   49   9   3   6   28.0  0.0  1.0   34   36.0  1.0  0.0  0.0   
996    75    4   14   2   0   1    1.0  0.0  1.0    7   17.0  0.0  0.0  0.0   
1545   15    0    2   1   0   0    2.0  0.0  0.0    1    2.0  1.0  0.0  0.0   
2015  630   68  163  36   0  22  116.0  0.0  1.0   67   85.0  8.0  2.0  2.0   
2285  686   70  182  39   3  15   95.0  2.0  4.0   48   70.0  9.0  2.0  2.0   
2952    0    1    0   0   0   0    0.0  0.0  0.0    0    0.0  0.0  0.0  0.0   
3085  947  106  211  29   0  39  159.0  8.0  3.0  122  167.0  1.0  1.0  0.0   

        SF  GIDP  
423    2.0   7.0  
996    0.0   3.0  
1545   0.0   2.0  
2015   4.0  22.0  
2285   6.0  17.0  
2952   0.0   0.0  
3085  16.0  26.0

In [49]:
# updating missing positions
no_pos = df_all[df_all.POS.isna()]
no_pos.POS = ['OF', 'OF', 'OF', 'OF','OF', '1B','1B']

df_all[df_all.POS.isna()] = no_pos

In [50]:
# Missing values reflect players who had no defensive statistics and no values present; to be replaced with zero
df_all = df_all.replace(np.nan, 0)

In [51]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2498 entries, 0 to 3394
Data columns (total 27 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   playerID    2498 non-null   object 
 1   GS          2498 non-null   float64
 2   InnOuts     2498 non-null   float64
 3   PO          2498 non-null   float64
 4   A           2498 non-null   float64
 5   E           2498 non-null   float64
 6   DP          2498 non-null   float64
 7   POS         2498 non-null   object 
 8   salary      2498 non-null   float64
 9   adj_salary  2498 non-null   float64
 10  G           2498 non-null   int64  
 11  AB          2498 non-null   int64  
 12  R           2498 non-null   int64  
 13  H           2498 non-null   int64  
 14  2B          2498 non-null   int64  
 15  3B          2498 non-null   int64  
 16  HR          2498 non-null   int64  
 17  RBI         2498 non-null   float64
 18  SB          2498 non-null   float64
 19  CS          2498 non-null  

- Great, no more missing values. For visual purposes the full names of players will be added and columns repositioned

In [52]:
# adding player names to dataframe 
name_df = names[['playerID', 'nameFirst', 'nameLast']]
name_df['name'] = name_df['nameFirst'] + ' ' + name_df['nameLast']
name_df = name_df[['playerID', 'name']]

In [53]:
name_df.head()

playerID           name
0  aardsda01  David Aardsma
1  aaronha01     Hank Aaron
2  aaronto01   Tommie Aaron
3   aasedo01       Don Aase
4   abadan01      Andy Abad

In [54]:
df_all = pd.merge(df_all, name_df, on=['playerID'], how='left')

In [55]:
df_all.head()

playerID     GS  InnOuts     PO       A     E     DP POS     salary  \
0   abadan01    4.0    138.0   37.0     1.0   1.0    3.0  1B   327000.0   
1  abbotje01  140.0   3688.0  299.0     2.0   8.0    0.0  OF   985000.0   
2  abbotku01  504.0  13474.0  938.0  1262.0  79.0  275.0  SS  4237000.0   
3  abercre01   96.0   2624.0  272.0     7.0   6.0    2.0  OF   327000.0   
4  abernbr01  217.0   5782.0  445.0   590.0  22.0  154.0  2B   515000.0   

   adj_salary    G    AB    R    H   2B  3B  HR    RBI    SB    CS   BB  \
0    528240.0   15    21    1    2    0   0   0    0.0   0.0   1.0    4   
1   2301147.0  233   596   82  157   33   2  18   83.0   6.0   5.0   38   
2  11970434.0  702  2044  273  523  109  23  62  242.0  22.0  11.0  133   
3    528240.0  180   386   65   86   20   2   9   34.0  18.0   8.0   21   
4    926963.0  232   868   97  212   36   5   8   79.0  21.0   7.0   60   

      SO   IBB   HBP    SH    SF  GIDP                name  
0    5.0   0.0   0.0   0.0   0.0   1.0           Andy Abad  
1   91.0   2.0   3.0   5.0   7.0  12.0         Jeff Abbott  
2  571.0  11.0  17.0  21.0  12.0  37.0         Kurt Abbott  
3  123.0   2.0   7.0   5.0   2.0   3.0  Reggie Abercrombie  
4   93.0   1.0   7.0  16.0   4.0  15.0     Brent Abernathy

In [56]:
def move_cols(df):
    
    '''
    update "columns" & "drops" lists before calling this function
    df will be reconfigured to show "columns" first and drop "drops"
    '''
    
    columns = ['name', 'salary', 'adj_salary', 'POS']   
    drops = ['playerID']
    
    df = df.copy()
    temp_df = df[columns].copy()
    df = df.drop(columns, axis=1)
    df = df.drop(drops, axis=1)
    new_df = pd.concat([temp_df, df], axis=1)
    
    return new_df

In [57]:
df_final = move_cols(df_all)

- A final look at the data then export to CSV

In [58]:
df_final.head()

name     salary  adj_salary POS     GS  InnOuts     PO  \
0           Andy Abad   327000.0    528240.0  1B    4.0    138.0   37.0   
1         Jeff Abbott   985000.0   2301147.0  OF  140.0   3688.0  299.0   
2         Kurt Abbott  4237000.0  11970434.0  SS  504.0  13474.0  938.0   
3  Reggie Abercrombie   327000.0    528240.0  OF   96.0   2624.0  272.0   
4     Brent Abernathy   515000.0    926963.0  2B  217.0   5782.0  445.0   

        A     E     DP    G    AB    R    H   2B  3B  HR    RBI    SB    CS  \
0     1.0   1.0    3.0   15    21    1    2    0   0   0    0.0   0.0   1.0   
1     2.0   8.0    0.0  233   596   82  157   33   2  18   83.0   6.0   5.0   
2  1262.0  79.0  275.0  702  2044  273  523  109  23  62  242.0  22.0  11.0   
3     7.0   6.0    2.0  180   386   65   86   20   2   9   34.0  18.0   8.0   
4   590.0  22.0  154.0  232   868   97  212   36   5   8   79.0  21.0   7.0   

    BB     SO   IBB   HBP    SH    SF  GIDP  
0    4    5.0   0.0   0.0   0.0   0.0   1.0  
1   38   91.0   2.0   3.0   5.0   7.0  12.0  
2  133  571.0  11.0  17.0  21.0  12.0  37.0  
3   21  123.0   2.0   7.0   5.0   2.0   3.0  
4   60   93.0   1.0   7.0  16.0   4.0  15.0

In [59]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2498 entries, 0 to 2497
Data columns (total 27 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        2498 non-null   object 
 1   salary      2498 non-null   float64
 2   adj_salary  2498 non-null   float64
 3   POS         2498 non-null   object 
 4   GS          2498 non-null   float64
 5   InnOuts     2498 non-null   float64
 6   PO          2498 non-null   float64
 7   A           2498 non-null   float64
 8   E           2498 non-null   float64
 9   DP          2498 non-null   float64
 10  G           2498 non-null   int64  
 11  AB          2498 non-null   int64  
 12  R           2498 non-null   int64  
 13  H           2498 non-null   int64  
 14  2B          2498 non-null   int64  
 15  3B          2498 non-null   int64  
 16  HR          2498 non-null   int64  
 17  RBI         2498 non-null   float64
 18  SB          2498 non-null   float64
 19  CS          2498 non-null  

In [60]:
df_final.to_csv('pt2_df.csv', index=False)

**END OF PART 1**